src/simulator/halo_simulator.py

```diff
@@ -223,17 +228,98 @@ def simulated_venn_diagram_reach_by_spend(
             estimate of the number of people reached in this subset.
         """
         raise NotImplementedError()
 
+    def _form_venn_diagram_regions(
+        self, spends: List[float], max_frequency: int = 1
+    ) -> Tuple[Dict[int, Set], Dict[int, List]]:
+        """Form Venn diagram regions that contain k+ reaches
+
+        For each subset of publishers, computes k+ reaches for those users
+        who are reached by the publishers in that subset.
+
+        Args:
+            spends:  The hypothetical spend vector, equal in length to
+              the number of publishers.  spends[i] is the amount that is
+              spent with publisher i. Note that publishers with 0 spends will
+              not be included in the Venn diagram reach.
+            max_frequency:  The maximum frequency for which to report reach.
+        Returns:
+            pub_set_by_region:  Contains the mapping of a binary representation
+              to a set of publisher ids that are in the corresponding region.
+            regions:  Contains k+ frequencies indexed by their binary
+              representation.
+        """
+        # Get user counts by spend for each active publisher
+        user_counts_by_pub_id = {}
+        for pub_id, spend in enumerate(spends):
+            if not spend:
+                continue
+            user_counts_by_pub_id[pub_id] = self._publishers[
+                pub_id
+            ]._publisher_data.user_counts_by_spend(spend)
+
+        if len(user_counts_by_pub_id) > MAX_ACTIVE_PUBLISHERS:
+            raise ValueError(
+                f"There are {len(user_counts_by_pub_id)} publishers"
+                f"for Venn diagram algorithm. The maximum limit is {MAX_ACTIVE_PUBLISHERS}"
+            )
+
+        # Locate user's region and sum the impressions. Also, map the region
+        # representation to publisher ids.
+        UserInfo = collections.namedtuple(
+            typename="Userinfo",
+            field_names="located_region impressions",
+            defaults=[0, 0],
+        )
+
+        users_info = collections.defaultdict(UserInfo)
+        pub_set_by_region = collections.defaultdict(set)
+
+        for pub_id, user_counts in user_counts_by_pub_id.items():
+            for user_id, impressions in user_counts.items():
+                lr = users_info[user_id].located_region
+                new_lr = lr | (1 << pub_id)
+
+                if not new_lr in pub_set_by_region:
+                    pub_set_by_region[new_lr] = copy.deepcopy(pub_set_by_region[lr])
+                    pub_set_by_region[new_lr].add(pub_id)
+
+                users_info[user_id] = UserInfo(
+                    new_lr, users_info[user_id].impressions + impressions
+                )
+
+        # Remove the null region
+        pub_set_by_region.pop(0)
+
+        # Fill in the occupied regions of the Venn diagram with capped user counts.
+        region_user_counts = collections.defaultdict(dict)
```

Comment:
I think lines 306-319 could be replaced with something similar to the following:

frequencies_by_region = [[0] * max_frequency for r in set(user_region.values())]
for user, region in user_region.items():
  frequencies_by_region[region][user_impressions[user]] += 1

... Now you need to convert this into a list of k+ reaches ...

Reply:
Thanks for the suggestion! Few suggestions on top of yours:
1. Change `frequencies_by_region` to be `Dict[int, List]`, so the indexing later would make sense.
2. Change the length of the lists in `frequencies_by_region` to (max_frequency + 1)
3. For `frequencies_by_region[region][user_impressions[user]] += 1`, `user_impressions[user]` might be larger than `max_frequency`. We can cap it beforehand, ie:
    ```python
     impressions = min(user_impressions[user], max_frequency)
    ```

Reply:
Done